File: ruby/problem12.rb
number = 0
i = 1

def number_of_divisors(number)
  divisors = 0
  sqrt = number**0.5
  (1..sqrt).each do |n|
    if number % n == 0
      divisors += 2
    end
  end
  divisors -=2 if sqrt*sqrt == number
  divisors
end

while number_of_divisors(number) < 500
  number += i
  i += 1
end

puts number

File: haskell/problem12.hs
isqrt :: Integer -> Integer
isqrt x = floor $ sqrt $ fromIntegral x
-- multiply length by two since the divisor has to be multiplied by something
divisors x = 2 * length [ y | y <- [1..isqrt x], x `mod` y == 0]

take 1 [ x | x <- scanl1 (+) [0..], divisors x > 500]
